### ISSUE TITLE
Make it more difficult to render a table without `widths` set

### ISSUE THREAD

**Issue by joshcbrown** [ratatui-org/ratatui#510]
## Problem
 
Currently, the `Table` widget initialises with an empty `widths` field. If the `widths` method isn't called before rendering, nothing is rendered in the table's allocated chunk. From an API standpoint, I find this confusing. I struggle to see a situation where not rendering any of the table's contents is desirable.

## Solution

I'm not sure that there's a way to implement this in a non-breaking way. The easiest thing to do would be to add a parameter for `widths` in the constructor of `Table`.

## Alternatives

I can think of a few other ways to implement my desired behaviour:
- Add a reasonable default for `widths`; something like each column taking up the same amount of room. I'm not sure this can be accomplished without changing `widths` to be an owned value.
- Perform an assertion before rendering that `widths` is non-empty.

If none of those changes are feasible given their breaking nature, it'd still be super nice to make the documentation clearer about the potential foot gun.

## Additional context
Nothing to add apart from I'd be very happy to implement this if approved.

**Comment by joshka** [comment]
The first option seems reasonable to me.
It breaks anyone intentionally omitting constraints and expecting that to render no columns visible but this doesn't seem like useful behavior to rely on.
Should probably be ~5 line fix in `Table::get_columns_widths()` something like:
```rust
let widths = if self.widths.is_empty() {
  let col_count = self.rows.iter().max_by(|r| r.cells.len());
  vec![Constraint::Ratio(1,col_count); col_count];
} else {
  self.widths.to_vec();
}
```

Plus unit tests, docs, etc. Do you want to make a PR for it?

**Comment by joshcbrown** [comment]
Thanks for the direction. Yes, happy to make a PR for it. Another way might be making `width` an `Option<&[Constraint]>`, which would allow users to opt in to rendering nothing. As you say though, it is a strange behaviour to rely on. What do you think?

**Comment by a-kenji** [comment]
You can always opt to render nothing, you don't need functionality on a custom widget for that.

**Comment by joshcbrown** [comment]
Yeah, cool. Will get to work on the PR.

**Comment by joshcbrown** [comment]
Ok, so I'm trying to push my changes to my fork, and the CI is failing on a clippy lint in `src/style/stylize.rs`. Not sure how to proceed - should I fix this in my commit?

**Comment by kdheepak** [comment]
( @joshka I'm not sure if I should open a separate issue for this or not. )

Right now there's no easy way to apply a custom width for each columns. I have to resort to using `.widths(&[Constraint::XXX(), ...])`. 

For example, I want to be able to calculate a layout and pass in widths for the table to use, so that I can add a keybinding for the user to increase the width of a particular column. The way to do it at the moment is do a layout calculation, get the widths, then again explicitly call `.widths(widths.iter().map(|w| Constraint::Length(w)).collect::Vec<Constraint>())` to generate the specific widths that I want.

It would be nice if we could rename the current `.widths` to `.width_constraints()` and have a new `.widths()` function that takes just a `width` directly.

**Comment by joshka** [comment]
TLDR - yes open an issue ;)

I'm a +1 that we need to nail down a better guarantee for table widths. In particular the gaps between columns should be the last constraint to be broken. I'd suggest taking a step back for that however and thinking in terms of design, use cases, specific tests and combinations that users might need for this. I think that using `taffy` based layout might be compelling to migrate towards rather than continuing to try to meet this on top of cassowary.

A rename like you're suggesting would break every application built on ratatui and prevent easily updating every tui app, so if we did this, it would probably be via a multi-step process (add a new method, deprecate the old, later delete the old method). But you might just avoid this for now by calling it something like `fixed_widths`.

The problem with that however is how to handle over allocation generically in a way that makes sense for all usages. I don't know the answer to that, so I'd want someone to put forward some considered approach and help show that it's the right one (compare it against options etc.)

**Comment by kdheepak** [comment]
I'll open an issue since it seems like it might be part of a larger discussion.

**Comment by pcgeek86** [comment]
This issue bit me as well. I created a table, with a few static rows / columns of data, and was getting blank output.

There was no feedback to indicate that I **must** call the `.widths()` method and pass in the column sizing constraints. Improving this usability would be helpful for new developers, like myself.

**Comment by joshka** [comment]
In https://github.com/ratatui-org/ratatui/pull/638 I've rewritten a bunch of docs to make it a bit more clear that widths is necessary.

I'd like to also suggest that we should introduce a breaking change to `Table::new()` so that the widths are a required parameter.

Summary of approaches:
- https://github.com/ratatui-org/ratatui/pull/519 - adds defaults if widths are not set
- https://github.com/ratatui-org/ratatui/pull/638 - adds documentation about setting .widths() being mandatory
- TODO breaking change to add widths to `Table::new()`
